src/net/server/channel/handlers/CouponCodeHandler.java

```diff
@@ -195,27 +195,40 @@ public final void handlePacket(SeekableLittleEndianAccessor slea, MapleClient c)
                 if (type < 0) {
                     c.announce(MaplePacketCreator.showCashShopMessage((byte) parseCouponResult(type)));
                 } else {
-                    List<Item> couponPackage = new LinkedList<>();
-
+                    List<Item> cashItems = new LinkedList<Item>();
+                    List<Pair<Integer, Integer>> items = new LinkedList<Pair<Integer, Integer>>();
```

Comment:
all usages of `LinkedList` should be `ArrayList`, as we're only adding to the end: adding to the end is `O(n)` in LinkedList and `O(1)` in ArrayList.

Reply:
`LinkedList`'s `add()` is equivalent to `addLast()`, which is O(1), whereas `ArrayList` has the potential of copying the array if the current capacity is exceeded.

Reply:
hmm, TIL!